File: src/types/rule/FocusRule_Subject.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "../Locale.js"

export const Subject = DB.Entity(import.meta.url, {
  name: "Subject",
  namePlural: "Subjects",
  comment: "Subjects or Subject Areas are the categories of Focus Rules.",
  type: () =>
    DB.Object({
      translations: NestedTranslationMap(
        DB.Required,
        "Subject",
        DB.Object({
          name: DB.Required({
            comment: "The subject.",
            type: DB.String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/Rule/FocusRule_Subject.swift
import FileDB

/// Subjects or Subject Areas are the categories of Focus Rules.
@Model
public struct Subject {
    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // SubjectTranslation
        /// The subject.
        @MinLength(1)
        let name: String
    }
}
